test/parallel/test-fs-write-file.js

```diff
@@ -95,3 +95,13 @@ fs.open(filename4, 'w+', common.mustSucceed((fd) => {
 
   process.nextTick(() => controller.abort());
 }
+
+{
+  // Test read-only mode
+  const filename3 = join(tmpdir.path, 'test3.txt');
+  const readOnlyOption = { mode: fs.constants.O_RDONLY, flag: 'r' };
+  fs.writeFile(filename3, s, readOnlyOption, common.expectsError({
+    code: 'EBADF',
+    message: 'EBADF: bad file descriptor, write'
```

Comment:
Not a change to make in this PR, but I wonder if that error message could be improved. Unless I'm misunderstanding, the problem from the user perspective is mode/permissions and not that the file descriptor is bad.

Reply:
Agreed. Since `writeFile(fd)` and `writeFile(path)` share the same util function `writeAll`, this change may need a refactor.

I'll open an issue to track it and look into it later.

Reply:
https://github.com/nodejs/node/issues/38607

Reply:
According to https://man7.org/linux/man-pages/man2/write.2.html:

* `EBADF`: fd is not a valid file descriptor or is not open for writing.
* `EPERM`: The operation was prevented by a file seal; see fcntl(2).

Considering that the file has been opened in the read-only mode and the program is attempting to write to it, isn't `EBADF` more appropriate here?

Reply:
This function just reflects the error that is returned from [`uv_fs_write`](http://docs.libuv.org/en/v1.x/fs.html#c.uv_fs_write), which again reflects the error the OS returns. I don't think we replace the error codes for the other functions, so doing it for this particular function feels a little odd, don't you think?

Reply:
@pd4d10 This test is failing for Windows. Could you please update it to expect an `EPERM` for Windows?